File: frontend/src/api/workStatus.ts
import apiClient from './client'

export interface WorkStatus {
  id: string
  employee_id: string
  status: 'ASSIGNED' | 'FREE_IMMEDIATE' | 'FREE_PLANNED' | 'INTERNAL' | 'LEAVE' | 'LEAVING'
  free_from: string | null
  note: string | null
  updated_at: string
}

export interface WorkStatusUpdateData {
  status: string
  free_from?: string | null
  note?: string | null
}

export interface AssignmentItem {
  id: string
  employee_id: string
  project_id: string
  project_name: string | null
  allocation_percent: number
  started_at: string
  ends_at: string | null
  is_active: boolean
  created_at: string
}

export interface AssignmentCreateData {
  project_id: string
  allocation_percent: number
  started_at: string
  ends_at?: string | null
}

export interface AssignmentUpdateData {
  allocation_percent?: number
  ends_at?: string | null
  is_active?: boolean
}

export const workStatusApi = {
  getWorkStatus: (employeeId: string) =>
    apiClient.get<WorkStatus>(`/employees/${employeeId}/work-status`).then(r => r.data),

  updateWorkStatus: (employeeId: string, data: WorkStatusUpdateData) =>
    apiClient.patch<WorkStatus>(`/employees/${employeeId}/work-status`, data).then(r => r.data),

  listAssignments: (employeeId: string) =>
    apiClient.get<AssignmentItem[]>(`/employees/${employeeId}/assignments`).then(r => r.data),

  createAssignment: (employeeId: string, data: AssignmentCreateData) =>
    apiClient.post<AssignmentItem>(`/employees/${employeeId}/assignments`, data).then(r => r.data),

  updateAssignment: (assignmentId: string, data: AssignmentUpdateData) =>
    apiClient.patch<AssignmentItem>(`/assignments/${assignmentId}`, data).then(r => r.data),
}

File: backend/app/models/work_status.py
"""
稼働状況・アサイン関連モデル

- WorkStatus: 稼働状況（1人1レコード）
- Assignment: アサイン管理
"""
import uuid
from datetime import date, datetime
from sqlalchemy import String, SmallInteger, Boolean, Text, Date, DateTime, Integer, ForeignKey, Enum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.core.database import Base

WorkStatusEnum = Enum(
    "ASSIGNED", "FREE_IMMEDIATE", "FREE_PLANNED", "INTERNAL", "LEAVE", "LEAVING",
    name="work_status",
)


class WorkStatus(Base):
    __tablename__ = "work_statuses"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    employee_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("employees.id", ondelete="CASCADE"), unique=True, nullable=False
    )
    status: Mapped[str] = mapped_column(WorkStatusEnum, nullable=False, default="FREE_IMMEDIATE")
    free_from: Mapped[date | None] = mapped_column(Date, nullable=True)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    updated_by: Mapped[str | None] = mapped_column(String(36), ForeignKey("employees.id"), nullable=True)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow
    )

    employee: Mapped["Employee"] = relationship("Employee", foreign_keys=[employee_id])
    updater: Mapped["Employee | None"] = relationship("Employee", foreign_keys=[updated_by])


class Assignment(Base):
    __tablename__ = "assignments"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    employee_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("employees.id", ondelete="CASCADE"), nullable=False
    )
    project_id: Mapped[str] = mapped_column(String(36), ForeignKey("projects.id"), nullable=False)
    allocation_percent: Mapped[int] = mapped_column(SmallInteger, nullable=False, default=100)
    started_at: Mapped[date] = mapped_column(Date, nullable=False)
    ends_at: Mapped[date | None] = mapped_column(Date, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    created_by: Mapped[str | None] = mapped_column(String(36), ForeignKey("employees.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)

    employee: Mapped["Employee"] = relationship("Employee", foreign_keys=[employee_id])
    project: Mapped["Project"] = relationship("Project", back_populates="assignments")
    creator: Mapped["Employee | None"] = relationship("Employee", foreign_keys=[created_by])


# 循環インポート回避
from app.models.employee import Employee  # noqa: E402
from app.models.project import Project  # noqa: E402
